client/lib/perfmon/index.js

```diff
@@ -24,128 +25,123 @@ const PLACEHOLDER_CLASSES = [
+	remove( activePlaceholders, node => ! OBSERVE_ROOT.contains( node ) || ! isPlaceholder( node ) );
 
-		// remove any nodes from activePlaceholders that are no longer placeholders
-		// check each node for:
-		// a. whether it's still in the DOM at all, and if so:
-		// b. whether it still has a placeholder class
-		remove( activePlaceholders, function( node ) {
-			return ! OBSERVE_ROOT.contains( node ) || ! isPlaceholder( node );
-		} );
+	checkActivePlaceholders();
+}
 
-		checkForVisiblePlaceholders( 'mutation' );
-	} );
+function createMutationObserver() {
+	const MutationObserver = window.MutationObserver || window.WebKitMutationObserver;
 
-	observer.observe( OBSERVE_ROOT, {
-		subtree: true,
-		attributes: true,
-		childList: true,
-		attributeFilter: [ 'class' ],
-	} );
+	mutationObserver = new MutationObserver( processMutations );
 }
 
-// check if there are any placeholders on the screen,
-// and trigger a timing event when all the placeholders are gone or the user
-// has navigated
-function checkForVisiblePlaceholders( trigger ) {
-	// determine how many placeholders are active in the viewport
-	const visibleCount = activePlaceholders.reduce( function( count, node ) {
-		return count + ( isElementVisibleInViewport( node ) ? 1 : 0 );
-	}, 0 );
-
-	// record event and reset timer if all placeholders are loaded OR user has just navigated
-	if ( placeholdersVisibleStart && ( visibleCount === 0 || trigger === 'navigate' ) ) {
-		// tell tracks to record duration
-		var duration = parseInt( performance.now() - placeholdersVisibleStart, 10 );
-		debug( `Recording placeholder wait. Duration: ${ duration }, Trigger: ${ trigger }` );
-		analytics.timing.record( `placeholder-wait`, duration, trigger );
-		placeholdersVisibleStart = visibleCount === 0 ? null : performance.now();
+// Start observing the DOM changes
+function startMutationObserver() {
+	if ( mutationObserver && ! mutationObserverActive ) {
+		mutationObserver.observe( OBSERVE_ROOT, OBSERVE_OPTIONS );
+		mutationObserverActive = true;
 	}
+}
 
-	// if we can see placeholders, placeholdersVisibleStart is falsy, start the clock
-	if ( visibleCount > 0 && ! placeholdersVisibleStart ) {
-		placeholdersVisibleStart = performance.now(); // TODO: performance.now()?
+// Stop observing the DOM changes and clean up all data structures
+function stopMutationObserver() {
+	if ( mutationObserver && mutationObserverActive ) {
+		activePlaceholders = [];
+		activePlaceholderEverDetected = false;
+		mutationObserver.disconnect();
+		mutationObserverActive = false;
 	}
+}
 
-	// if there are placeholders hanging around, print some useful stats
-	if ( activePlaceholders.length > 0 ) {
-		debug( 'Active placeholders: ' + activePlaceholders.length );
-		debug( 'Visible in viewport: ' + visibleCount );
+// Determine what kind of timing are we recording.
+// Initial app load or after-load navigation to another route?
+function determineTimingType() {
+	if ( navigationCount > 1 ) {
```

Comment:
Do we ever reset `navigationCount` to `0` after customer navigates away from a page where perfmon is enabled?

Reply:
If the customer navigates away from Calypso, the web page is destroyed, including the `navigationCount` variable. When you come back later, it gets initialized to 0 again.

After the customer navigates from `/stats` to `/posts` for example, the `navigationCount` is incremented. It's a navigation inside an already loaded app. That's what we want to detect. On initial load of Calypso, `navigationCount` gets incremented to 1 by the first `page('/route')` handler. We handle this load as `initial`. After the next navigation, it gets incremented to 2 and more, and we handle that as `navigation`.

Reply:
Ah, I see. Wouldn't it make more sense if it was a boolean instead of a number? Like `initialLoad`. Once set to `true`, it would log the navigation only.

Reply:
Yes, `initialLoad` would work, too. But I don't know how to set it. Before the first `page` call, it's `true`, after the first call, it's still `true`, and only after the second call it becomes `false`. I would still need to maintain a counter.